[cxx-interop] Support C++ default arguments

egorzhdan opened the pull request:
This allows calling a C++ function with default arguments from Swift without having to explicitly specify the values of all arguments.

For every C++ default argument we mark the corresponding Swift argument as defaulted (`= cxxDefaultArg`). If the default value of the argument is used, the compiler creates a new C++ function `__defaultArg_{argument index}_{mangled name of the function}` that returns the default expression. This function is then imported into Swift and invoked from a default expression of the imported function on the Swift side.

This patch doesn't cover default arguments of C++ constructors (https://github.com/apple/swift/issues/70124).

rdar://103975014

egorzhdan committed:
[cxx-interop] Support C++ default arguments

This allows calling a C++ function with default arguments from Swift without having to explicitly specify the values of all arguments.

rdar://103975014

hyp commented:
Could you also test things that get synthesized, e.g. base members in derived classes.

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@hyp good idea, I'll add tests for that

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please test source compatibility

egorzhdan commented:
@swift-ci please test Windows

rjmccall commented:
Haven't looked at the patch yet, but do we need to have constraints here to ensure that lifetimes aren't messed up?  Like, what happens if the parameter type is a reference?

hyp commented on test/Interop/Cxx/function/Inputs/default-arguments.h:
```diff
@@ -0,0 +1,55 @@
+  bool isZero(int v = 0) const { return v == 0; }
+};
+
+struct DerivedFromHasMethodWithDefaultArg : public HasMethodWithDefaultArg {};
```
Can you add a test for another level of derived (DerivedDerived) as there's another synthesis interaction happening past the first derived type.

egorzhdan replied:
Done

hyp commented on test/Interop/Cxx/function/default-arguments-module-interface.swift:
```diff
@@ -0,0 +1,31 @@
+// RUN: %target-swift-ide-test -print-module -module-to-print=DefaultArguments -I %S/Inputs -source-filename=x -cxx-interoperability-mode=upcoming-swift | %FileCheck %s
+
+// CHECK: func isZero(_ value: Int32 = cxxDefaultArg) -> Bool
```
is it possible to map some default arguments to literal constants instead, to present a better module interface?

egorzhdan replied:
Good point, I was initially thinking of reusing the parsing logic in `ImportMacro.cpp` to import simple default values such as `nullptr` without synthesizing new functions. I didn't have time to implement that, but I think this would be a great future improvement.

hyp commented on test/Interop/Cxx/function/Inputs/default-arguments.h:
```diff
@@ -0,0 +1,55 @@
+inline int ambiguous(int a) { return a; }
+
+inline int nonTrailing(int a, int b = 2) { return a + b; }
+inline int nonTrailing(int a = 1, int b);
```
Can you add some additional tests for a class template, including a dependent value in a default parameter?

egorzhdan replied:
Done

egorzhdan commented:
Thanks @rjmccall!

> Like, what happens if the parameter type is a reference?

We shouldn't add a default value for such parameter in Swift, since the parameter would get imported as `inout` and Swift doesn't allow default values for `inout`. This is now fixed.

I also added a test to make sure we don't add default values for foreign reference type params, we can make them work later.

Another interesting case is if a parameter has a pointer type. I think it's reasonable to assume that if the default expression allocates something, it will get deallocated by the callee. In that case we don't need to worry about lifetimes. In general, it seems importing default values of pointer parameters doesn't add extra unsafety, since the default value would refer to either a new pointer created inline or to something static/immortal.

hyp commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2478,6 +2479,19 @@ static ParamDecl *getParameterInfo(ClangImporter::Implementation *impl,
+            param, swiftParamTy, paramInfo->getParameterNameLoc())) {
+      paramInfo->setDefaultArgumentKind(DefaultArgumentKind::Normal);
+      paramInfo->setDefaultExpr(defaultArgExpr, /*isTypeChecked*/ true);
+      paramInfo->setDefaultValueStringRepresentation("cxxDefaultArg");
```
Another idea - since using a specific string value to represent the default value in the module interface, can you instead render the original Clang expression and present it to the user as a sort of special expression , e.g. `foo(x: Foo = #cxxDefaultArg(Foo(1))`

egorzhdan replied:
That is smart, I like this idea!

hyp commented:
Also, please check the source compatibility test suite as well :)

rjmccall commented:
> We shouldn't add a default value for such parameter in Swift, since the parameter would get imported as `inout` and Swift doesn't allow default values for `inout`. This is now fixed.

What if it's a `const &` parameter?  We don't import those as `inout`, right?

> Another interesting case is if a parameter has a pointer type. I think it's reasonable to assume that if the default expression allocates something, it will get deallocated by the callee. In that case we don't need to worry about lifetimes. In general, it seems importing default values of pointer parameters doesn't add extra unsafety, since the default value would refer to either a new pointer created inline or to something static/immortal.

I agree that pointers are probably relatively unlikely to be a problem, but view types might be dangerous, e.g. a `std::string_view` that's constructed from a temporary `std::string`.  But we should be able to detect this condition somewhat precisely, right?  The Clang AST has the actual default argument expression, and we can just look at that expression to see if it creates any temporaries and then say that such arguments can't be imported as default arguments.  (We can ignore the top-level temporary for the argument itself if the argument is a pr-value.)  This could be over-restrictive, e.g. if the default argument creates temporaries but the program doesn't depend on them staying alive across the call, but it's conservatively correct.  And it would be possible to lift that restriction in the future if we can figure out some way to extend temporaries across the call.

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please test source compatibility

egorzhdan commented:
@swift-ci please test source compatibility debug

rjmccall commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2484,6 +2486,73 @@ static ParamDecl *getParameterInfo(ClangImporter::Implementation *impl,
+            isSafeToImport = false;
+        }
+      }
+    }
```
Can all this reasonably just be a function?

rjmccall commented on lib/ClangImporter/ImporterImpl.h:
```diff
@@ -516,7 +516,7 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
 
   /// Clang parser, which is used to load textual headers.
-  std::unique_ptr<clang::MangleContext> Mangler;
+  std::shared_ptr<clang::MangleContext> Mangler;
```
Why is this necessary?

egorzhdan replied:
I reverted this to `unique_ptr` and replaced with something simpler.

We're now reusing `ClangImporter::Implementation::getMangledName` to create an Itanium-mangled name for the synthesized Swift function that generates the default argument. We can't reuse `ClangImporter::Implementation::Mangler` though, because it might not be an Itanium mangler, which would produce a name that isn't a valid identifier in Swift.

rjmccall commented on lib/IRGen/GenClangDecl.cpp:
```diff
@@ -54,6 +59,29 @@ class ClangDeclFinder
+          callback(calledClangFuncDecl);
+        }
+      }
+    }
```
So if we ever reference a function, we have to eagerly emit default argument generators for all of its parameters?  Is there some more reasonable way to make SIL carry the information for how to generate the body here so that we can just trigger this when we have a reference to the default argument generator?

egorzhdan replied:
We don't actually emit the generator unless it is used. Clang's CodeGen adds this decl to the list of deferred decls (https://github.com/apple/llvm-project/blob/d4738f0f52791c6004131af7322f13a07abee8e7/clang/lib/CodeGen/CodeGenModule.cpp#L3679C2-L3679C2) and emits it only if it's referenced.

I added an IRGen test to make sure unused generators aren't emitted.

rjmccall replied:
Ah, that makes sense.

rjmccall commented on lib/SIL/IR/SILDeclRef.cpp:
```diff
@@ -981,6 +981,10 @@ bool SILDeclRef::isForeignToNativeThunk() const {
+  // A default argument generator for a C++ function is a Swift function, so no
+  // thunk needed.
+  if (isDefaultArgGenerator())
+    return false;
```
So why is it marked as foreign?

egorzhdan replied:
The parent function is marked as foreign because it has a Clang node, but the default argument generator isn't foreign, and doesn't require a thunk despite its parent function being foreign. This check needs to happen before the `requiresForeignToNativeThunk` check below causes us to assume that a thunk is needed.

Oudinunit replied:
Hello How are you @egorzhdan @rjmccall @hyp @Oudinunit @lilyball @drodriguez @drodriguez @mattt @xwu @jtbandes @erica

rjmccall commented on lib/SILGen/SILGen.cpp:
```diff
@@ -719,6 +719,26 @@ static bool isEmittedOnDemand(SILModule &M, SILDeclRef constant) {
   case SILDeclRef::Kind::EnumElement:
     return true;
+  case SILDeclRef::Kind::DefaultArgGenerator: {
+    // Default arguments of C++ functions are only emitted is used.
```
```suggestion
    // Default arguments of C++ functions are only emitted if used.
```

rjmccall commented on lib/SILGen/SILGen.cpp:
```diff
@@ -719,6 +719,26 @@ static bool isEmittedOnDemand(SILModule &M, SILDeclRef constant) {
+      // a new Swift function that dispatches to the base method and has no
+      // clang node: the underlying clang decl is synthesized lazily. This means
+      // we can't rely on `funcDecl->hasClangNode()` here. Instead, check if any
+      // of the parameters has a clang node.
```
That feels like something we should record about the synthesized Swift function.

egorzhdan replied:
Oh, I didn't realize we could just check the function's module context. Fixed.

Oudinunit commented:
Hello I am Nittaya Schlamp Nice to meet you my family ![image](https://github.com/apple/swift/assets/107595400/efcebab6-6a80-42d5-a1aa-f6e4c68aa941)

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please smoke test

rjmccall reviewed (approved):
LGTM

Oudinunit commented:
I see you guys controlling me. Because I saw my profile overlapping 1 profile on GitHub. in my phone. I normally only have 2 profiles. But I have 3 profiles now. Because I don’t have profile me. again. you  know my family @egorzhdan
